### PR TITLE
EASY-2342: Request met lege authentication credentials resulteert in 503

### PR DESCRIPTION
**Pull request by vesaakerman**
Fixes EASY-2342

#### When applied it will
* return `401` when `user id` is empty or `password` is empty

#### How should this be manually tested?
See the instructions in teh Jira issue

#### Related pull requests on github
repo                       | PR                | note
-------------------------- | ----------------- | ----
easy-                      | [PR#](PRlink)     | ..

### CODE REVIEW

File: src/main/scala/nl.knaw.dans.easy.download/components/AuthenticationComponent.scala [hunk 1]
```diff
@@ -69,6 +69,8 @@ trait AuthenticationComponent extends DebugEnhancedLogging {
           .tried
       }.recoverWith {
         case t: AuthenticationException => Failure(InvalidUserPasswordException(userName, new Exception("invalid password", t)))
+        case t: InvalidNameException => Failure(InvalidNameAuthenticationException(t))
```
Comment (jo-pol) [inline src/main/scala/nl.knaw.dans.easy.download/components/AuthenticationComponent.scala]: No need for three different exceptions. Use something like 

    case t: InvalidNameException | InvalidNameException => Failure(InvalidUserPasswordException(...))

It also reduces the pattern match in `EasyDownloadServlet`

File: src/main/scala/nl.knaw.dans.easy.download/components/AuthenticationComponent.scala [hunk 2]
```diff
@@ -69,6 +69,8 @@ trait AuthenticationComponent extends DebugEnhancedLogging {
...
+        case t: OperationNotSupportedException if (password.isEmpty) => Failure(NoPasswordAuthenticationException(t))
```
Comment (jo-pol) [inline src/main/scala/nl.knaw.dans.easy.download/components/AuthenticationComponent.scala]: Shouldn't this become an internal error? It looks an implementation problem to me.

File: src/main/scala/nl.knaw.dans.easy.download/components/AuthenticationComponent.scala (hunk 1 above)
Comment (vesaakerman) [reply src/main/scala/nl.knaw.dans.easy.download/components/AuthenticationComponent.scala]: When you use `InvalidUserPasswordException`, you will not see the `cause` of the failure.

In InvalidUserPasswordException:
      `logger.info(s"invalid credentials for $userName: ${ cause.getMessage }")`

In the case of `InvalidNameException` and `OperationNotSupportedException` it is better to log the `cause` of the exception.

File: src/main/scala/nl.knaw.dans.easy.download/components/AuthenticationComponent.scala (hunk 2 above)
Comment (vesaakerman) [reply src/main/scala/nl.knaw.dans.easy.download/components/AuthenticationComponent.scala]: No, this in not internal error.  `Ldap` returns `InvalidNameException` when `user-id is empty`. And it returns `OperationNotSupportedException` when `password is empty`.

File: src/main/scala/nl.knaw.dans.easy.download/components/AuthenticationComponent.scala (hunk 1 above)
Comment (jo-pol) [reply src/main/scala/nl.knaw.dans.easy.download/components/AuthenticationComponent.scala]: By the way, the error message says "invalid password". 

Response to the end user should never tell whether the user name or password was invalid, just that the combination is invalid. An invalid password might tell a hacker he guessed the user right. Internally that difference might be important whenever we decide to add a policy to increase response time for a user on an invalid password. Though I rather think that should be an ldap feature than an application feature.

Comment (vesaakerman) [reply src/main/scala/nl.knaw.dans.easy.download/components/AuthenticationComponent.scala]: The user gets in these situations also only a message that he is unauthorized.  The more detailed cause comes only in the log file.

**Pull request merged**